src/libcore/num/dec2flt/rawfp.rs

```diff
@@ -0,0 +1,356 @@
+pub fn encode_normal<T: RawFloat>(x: Unpacked) -> T {
+    debug_assert!(T::min_sig() <= x.sig && x.sig <= T::max_sig(),
+        "encode_normal: significand not normalized");
+    // Remove the hidden bit
+    let sig_enc = x.sig & !(1 << T::explicit_sig_bits());
+    // Adjust the exponent for exponent bias and mantissa shift
+    let k_enc = x.k + T::max_exp() + T::explicit_sig_bits() as i16;
+    debug_assert!(k_enc != 0 && k_enc < T::max_encoded_exp(),
+        "encode_normal: exponent out of range");
+    // Leave sign bit at 0 ("+"), our numbers are all positive
+    let bits = (k_enc as u64) << T::explicit_sig_bits() | sig_enc;
+    T::from_bits(bits)
+}
+
+/// Construct the subnormal. A mantissa of 0 is allowed and constructs zero.
+pub fn encode_subnormal<T: RawFloat>(significand: u64) -> T {
+    assert!(significand < T::min_sig(), "encode_subnormal: not actually subnormal");
+    // Êncoded exponent is 0, the sign bit is 0, so we just have to reinterpret the bits.
+    T::from_bits(significand)
+}
+
+/// Approximate a bignum with an Fp. Rounds within 0.5 ULP with half-to-even.
+pub fn big_to_fp(f: &Big) -> Fp {
+    let end = f.bit_length();
+    assert!(end != 0, "big_to_fp: unexpectedly, input is zero");
+    let start = end.saturating_sub(64);
+    let leading = num::get_bits(f, start, end);
+    // We cut off all bits prior to the index `start`, i.e., we effectively right-shift by
+    // an amount of `start`, so this is also the exponent we need.
+    let e = start as i16;
+    let rounded_down = Fp { f: leading, e: e }.normalize();
+    // Round (half-to-even) depending on the truncated bits.
+    match num::compare_with_half_ulp(f, start) {
+        Less => rounded_down,
+        Equal if leading % 2 == 0 => rounded_down,
+        Equal | Greater => match leading.checked_add(1) {
+            Some(f) => Fp { f: f, e: e }.normalize(),
+            None => Fp { f: 1 << 63, e: e + 1 },
+        }
+    }
+}
+
+/// Find the largest floating point number strictly smaller than the argument.
+/// Does not handle subnormals, zero, or exponent underflow.
+pub fn prev_float<T: RawFloat>(x: T) -> T {
+    match x.classify() {
+        Infinite => panic!("prev_float: argument is infinite"),
+        Nan => panic!("prev_float: argument is NaN"),
+        Subnormal => panic!("prev_float: argument is subnormal"),
+        Zero => panic!("prev_float: argument is zero"),
+        Normal => {
+            let Unpacked { sig, k } = x.unpack();
+            if sig == T::min_sig() {
+                encode_normal(Unpacked::new(T::max_sig(), k - 1))
+            } else {
+                encode_normal(Unpacked::new(sig - 1, k))
+            }
+        }
+    }
+}
+
+// Find the smallest floating point number strictly larger than the argument.
+// This operation is saturating, i.e. next_float(inf) == inf.
+// Unlike most code in this module, this function does handle zero, subnormals, and infinities.
```

Comment:
This should explicitly document that it only handles floats with a positive sign bit.